core/src/main/java/org/sonarsource/sonarlint/core/SonarLintClientImpl.java

```diff
@@ -45,100 +47,247 @@
-    } catch (RuntimeException e) {
-      throw handleException(e);
+      checkStarted();
+      return globalContainer;
+    } finally {
+      rwl.readLock().unlock();
     }
-    this.started = true;
   }
 
   @Override
-  public RuleDetails getRuleDetails(String ruleKey) {
-    checkStarted();
-    return globalContainer.getRuleDetails(ruleKey);
+  public synchronized void start() {
+    rwl.writeLock().lock();
+    try {
+      if (started) {
+        throw new IllegalStateException("SonarLint Engine is already started");
+      }
+      this.globalContainer = GlobalContainer.create(globalConfig);
+      try {
+        globalContainer.startComponents();
+      } catch (RuntimeException e) {
+        throw SonarLintWrappedException.build(e);
+      }
+      this.started = true;
+    } finally {
+      rwl.writeLock().unlock();
+    }
   }
 
   @Override
-  public Collection<String> getActiveRuleKeys() {
-    checkStarted();
-    return globalContainer.getActiveRuleKeys();
+  public RuleDetails getRuleDetails(String ruleKey) {
+    rwl.readLock().lock();
+    try {
+      checkStarted();
+      return globalContainer.getRuleDetails(ruleKey);
+    } finally {
+      rwl.readLock().unlock();
+    }
   }
 
   @Override
   public AnalysisResults analyze(AnalysisConfiguration configuration, IssueListener issueListener) {
     checkNotNull(configuration);
-    Preconditions.checkNotNull(issueListener);
-    checkStarted();
+    checkNotNull(issueListener);
+    rwl.readLock().lock();
     try {
-      return globalContainer.analyze(configuration, issueListener);
-    } catch (RuntimeException e) {
-      throw handleException(e);
+      checkStarted();
+      try {
+        return globalContainer.analyze(configuration, issueListener);
+      } catch (RuntimeException e) {
+        throw SonarLintWrappedException.build(e);
+      }
+    } finally {
+      rwl.readLock().unlock();
     }
   }
 
-  private void checkStarted() {
-    if (!started) {
-      throw new IllegalStateException("SonarLint Engine is not started");
+  @Override
+  public GlobalSyncStatus getSyncStatus() {
+    checkMode();
+    rwl.readLock().lock();
+    try {
+      checkStarted();
+      return ((StorageGlobalContainer) globalContainer).getSyncStatus();
+    } finally {
+      rwl.readLock().unlock();
+    }
+  }
+
+  @Override
+  public void sync(ServerConfiguration serverConfig) {
+    checkNotNull(serverConfig);
+    checkMode();
+    rwl.writeLock().lock();
+    try {
+      checkStopped();
+      ConnectedContainer connectedContainer = new ConnectedContainer(globalConfig, serverConfig);
+      try {
+        connectedContainer.startComponents();
+      } catch (RuntimeException e) {
+        throw SonarLintWrappedException.build(e);
+      }
+      try {
+        connectedContainer.sync();
+      } finally {
+        try {
+          connectedContainer.stopComponents(false);
+        } catch (RuntimeException e) {
+          throw SonarLintWrappedException.build(e);
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Remove this throw statement from this finally block. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1143)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not throw exceptions in finally blocks. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1163)